### PR TITLE
Port the backend of FractionalMaxPool3d from TH to ATen

### PR DESCRIPTION
**Pull request by chandlerzuo**
1. Port the FractionalMaxPool3d implementation from THNN/THCUNN to ATen.
2. Expose this function to Python module nn.

### CODE REVIEW

**Comment by ezyang** [comment]
This PR is going to be in a similar situation to #15531 so I am going to hold off on reviewing it until we resolve the issues there.

File: test/common_nn.py
Comment (ezyang) [inline test/common_nn.py]: First off, it's awesome that you added these tests, thanks a lot :) (I didn't even realize that this function wasn't exposed from Python when I originally bootcamped this task). What would be really good is to add some tests which test for asymmetric kernel size. Something like in https://github.com/pytorch/pytorch/pull/15823

File: aten/src/ATen/native/FractionalMaxPool3d.cpp
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool3d.cpp]: Ordering here will need fixing (a test that takes tuple pool size will help you observe the problem)

File: aten/src/ATen/native/FractionalMaxPool3d.cpp
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool3d.cpp]: will need reordering (also below)

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: It looks like we lost the specializations for pool size 2-7, although I am not sure we actually care.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: heh, this is an even weirder ordering to have picked! It should be T (0), H (1), W (2)

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: you haven't fixed this one

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: and these need fixing too

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Missing cudaGetLastError check

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: unnecessary resize

File: torch/nn/functional.py
Comment (ezyang) [inline torch/nn/functional.py]: Once we reorder the parameters, the docs here should change to give the conventional ordering TxHxW

**Comment by ezyang** [comment]
I finished a complete review; I think my comments are the only outstanding issues.

File: aten/src/ATen/native/FractionalMaxPool3d.cpp
Comment (chandlerzuo) [reply aten/src/ATen/native/FractionalMaxPool3d.cpp]: Fixed

File: aten/src/ATen/native/FractionalMaxPool3d.cpp
Comment (chandlerzuo) [reply aten/src/ATen/native/FractionalMaxPool3d.cpp]: I have changed all int back to int64_t. I think it is very unlikely to have these variables overflow for real data, but agree that we need to make this error prone.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: It is still preserved in the fractional_max_pool3d_out_frame function.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: aten/src/ATen/native/cuda/FractionalMaxPool3d.cu
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool3d.cu]: Fixed.

File: test/common_nn.py
Comment (chandlerzuo) [reply test/common_nn.py]: The asymmetric test is added as suggested.

**Review by ezyang — APPROVED** [review]
Make sure CI passes before landing

**Pull request closed without merging**
